File: native/cocos/renderer/renderer/Light.cpp
/****************************************************************************
 Copyright (c) 2018 Xiamen Yaji Software Co., Ltd.

 http://www.cocos2d-x.org
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/

#include "Light.h"
#include <new>
#include "gfx/DeviceGraphics.h"
#include "gfx/Texture2D.h"
#include "gfx/RenderBuffer.h"
#include "gfx/FrameBuffer.h"

RENDERER_BEGIN

Light::Light()
{
}

void Light::setColor(float r, float g, float b)
{
    _color.set(r, g, b);
    _colorUniform.set(r * _intensity, g * _intensity, b * _intensity);
}

void Light::setIntensity(float val)
{
    _intensity = val;
    _colorUniform.set(_color.r * _intensity,
                      _color.g * _intensity,
                      _color.b * _intensity);
}

void Light::setSpotAngle(float val)
{
    _spotAngle = val;
    _spotUniform[0] = std::cos(_spotAngle * 0.5f);
}

void Light::setSpotExp(float val)
{
    _spotExp = val;
    _spotUniform[1] = _spotExp;
}

void Light::setShadowType(ShadowType val)
{
    if (ShadowType::NONE == _shadowType &&
        ShadowType::NONE != val)
    {
        _shadowMapDirty = true;
    }
    
    _shadowType = val;
}

void Light::setShadowResolution(uint32_t val)
{
    if (_shadowResolution != val)
    {
        _shadowMapDirty = true;
        _shadowResolution = val;
    }
}

float Light::getShadowMinDepth() const
{
    if (LightType::DIRECTIONAL == _type)
        return 1.0f;
    else
        return _shadowMinDepth;
}

float Light::getShadowMaxDepth() const
{
    if (LightType::DIRECTIONAL == _type)
        return 1.0f;
    else
        return _shadowMaxDepth;
}

void Light::setWorldMatrix(const Mat4& worldMatrix)
{
    _worldMatrix = worldMatrix;
    Quaternion rotation;
    Vec3 translation;
    _worldMatrix.decompose(nullptr, &rotation, &translation);
    Mat4::createRotation(rotation, &_worldRT);
    _worldRT.translate(translation);
}

void Light::extractView(View& out, const std::vector<std::string>& stages)
{
    out.shadowLight = const_cast<Light*>(this);
    out.rect = { 0, 0, (float)_shadowResolution, (float)_shadowResolution };
    
    // clear options
    out.color.set(1.f, 1.f, 1.f, 1.f);
    out.depth = 1;
    out.stencil = 1;
    out.clearFlags = ClearFlag::COLOR | ClearFlag::DEPTH;
    
    // stage & framebuffer
    out.stages = stages;
    out.frameBuffer = _shadowFrameBuffer;
    
    // view projection matrix
    switch (_type)
    {
        case LightType::SPOT:
            computeSpotLightViewProjMatrix(out.matView, out.matProj);
            break;
        case LightType::DIRECTIONAL:
            computeDirectionalLightViewProjMatrix(out.matView, out.matProj);
            break;
        case LightType::POINT:
            computePointLightViewProjMatrix(out.matView, out.matProj);
            break;
        default:
            RENDERER_LOGW("Shadow of this light type is not supported");
            break;
    }
    
    // view-porjection
    Mat4::multiply(out.matView, out.matProj, &out.matViewProj);
    out.matInvViewPorj = out.matViewProj.getInversed();
    
    _viewProjMatrix.set(out.matViewProj);
}

void Light::update(DeviceGraphics* device)
{
    updateLightPositionAndDirection();
    
    if (ShadowType::NONE == _shadowType)
        destroyShadowMap();
    else
    {
        destroyShadowMap();
        generateShadowMap(device);
        _shadowMapDirty = false;
    }
}

// private functions

void Light::updateLightPositionAndDirection()
{
    _worldMatrix.transformVector(Vec3(0, 0, -1.f), &_directionUniform);
    _positionUniform.set(_worldMatrix.m[12], _worldMatrix.m[13], _worldMatrix.m[14]);
}

void Light::generateShadowMap(DeviceGraphics* device)
{
    _shadowMap = new (std::nothrow) Texture2D();
    if (!_shadowMap)
        return;
    
    _shadowDepthBuffer = new (std::nothrow) RenderBuffer();
    if (!_shadowDepthBuffer)
    {
        _shadowMap->release();
        _shadowMap = nullptr;
        
        return;
    }
    
    _shadowFrameBuffer = new (std::nothrow) FrameBuffer();
    if (!_shadowFrameBuffer)
    {
        _shadowMap->release();
        _shadowDepthBuffer->release();
        _shadowMap = nullptr;
        _shadowDepthBuffer = nullptr;
        
        return;
    }
    
    Texture2D::Options options;
    options.width = _shadowResolution;
    options.height = _shadowResolution;
    options.format = Texture2D::Format::RGBA8;
    options.wrapS = Texture2D::WrapMode::CLAMP;
    options.wrapT = Texture2D::WrapMode::CLAMP;
    _shadowMap->init(device, options);
    
    _shadowDepthBuffer->init(device, RenderBuffer::Format::D16, _shadowResolution, _shadowResolution);
    
    _shadowFrameBuffer->init(device, _shadowResolution, _shadowResolution);
    _shadowFrameBuffer->setColorBuffer(_shadowMap, 0);
    _shadowFrameBuffer->setDepthBuffer(_shadowDepthBuffer);
}

void Light::destroyShadowMap()
{
    if (_shadowMap)
    {
        _shadowMap->release();
        _shadowMap = nullptr;
        
        RENDERER_SAFE_RELEASE(_shadowDepthBuffer);
        _shadowDepthBuffer = nullptr;
        
        RENDERER_SAFE_RELEASE(_shadowFrameBuffer);
        _shadowFrameBuffer = nullptr;
    }
}

void Light::computeSpotLightViewProjMatrix(Mat4& matView, Mat4& matProj) const
{
    // view matrix
    matView = _worldRT.getInversed();
    
    // proj matrix
    Mat4::createPerspective(_spotAngle * _spotAngleScale,
                            1,
                            _shadowMinDepth,
                            _shadowMaxDepth,
                            &matProj);
}

void Light::computeDirectionalLightViewProjMatrix(Mat4& matView, Mat4& matProj) const
{
    // view matrix
    matView = _worldRT.getInversed();
    
    // proj matrix
    uint32_t halfSize = _shadowFustumSize / 2;
    Mat4::createOrthographic(-halfSize, halfSize, -halfSize, halfSize, &matProj);
}

void Light::computePointLightViewProjMatrix(Mat4& matView, Mat4& matProj) const
{
    //TODO
}

RENDERER_END

File: cocos/core/renderer/scene/light.ts
/*
 Copyright (c) 2020 Xiamen Yaji Software Co., Ltd.

 https://www.cocos.com/

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated engine source code (the "Software"), a limited,
 worldwide, royalty-free, non-assignable, revocable and non-exclusive license
 to use Cocos Creator solely to develop games on your target platforms. You shall
 not use Cocos Creator software for developing other software or tools that's
 used for developing games. You are not granted to publish, distribute,
 sublicense, and/or sell copies of Cocos Creator.

 The software or tools in this License Agreement are licensed, not sold.
 Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 */

import { JSB } from 'internal:constants';
import { Vec3 } from '../../math';
import { TransformBit } from '../../scene-graph/node-enum';
import { RenderScene } from './render-scene';
import { Node } from '../../scene-graph';
import { NativeDirectionalLight, NativeLight, NativeSphereLight, NativeSpotLight } from './native-scene';

// Color temperature (in Kelvin) to RGB
export function ColorTemperatureToRGB (rgb: Vec3, kelvin: number) {
    if (kelvin < 1000.0) {
        kelvin = 1000.0;
    } else if (kelvin > 15000.0) {
        kelvin = 15000.0;
    }

    // Approximate Planckian locus in CIE 1960 UCS
    const kSqr = kelvin * kelvin;
    const u = (0.860117757 + 1.54118254e-4 * kelvin + 1.28641212e-7 * kSqr) / (1.0 + 8.42420235e-4 * kelvin + 7.08145163e-7 * kSqr);
    const v = (0.317398726 + 4.22806245e-5 * kelvin + 4.20481691e-8 * kSqr) / (1.0 - 2.89741816e-5 * kelvin + 1.61456053e-7 * kSqr);

    const d = (2.0 * u - 8.0 * v + 4.0);
    const x = (3.0 * u) / d;
    const y = (2.0 * v) / d;
    const z = (1.0 - x) - y;

    const X = (1.0 / y) * x;
    const Z = (1.0 / y) * z;

    // XYZ to RGB with BT.709 primaries
    rgb.x =  3.2404542 * X + -1.5371385 + -0.4985314 * Z;
    rgb.y = -0.9692660 * X +  1.8760108 +  0.0415560 * Z;
    rgb.z =  0.0556434 * X + -0.2040259 +  1.0572252 * Z;
}

export enum LightType {
    DIRECTIONAL,
    SPHERE,
    SPOT,
    UNKNOWN,
}

export const nt2lm = (size: number) => 4 * Math.PI * Math.PI * size * size;

export class Light {
    protected declare _nativeObj: NativeLight | null;
    protected _init (): void {
        if (JSB) {
            switch (this._type) {
            case LightType.DIRECTIONAL:
                this._nativeObj = new NativeDirectionalLight();
                break;
            case LightType.SPHERE:
                this._nativeObj = new NativeSphereLight();
                break;
            case LightType.SPOT:
                this._nativeObj = new NativeSpotLight();
                break;
            default:
                break;
            }
            this._nativeObj!.setType(this._type);
        }
    }
    protected _destroy (): void {
        if (JSB) {
            this._nativeObj = null;
        }
    }

    get baked () {
        return this._baked;
    }

    set baked (val) {
        this._baked = val;
        if (JSB) {
            this._nativeObj!.setBaked(val);
        }
    }

    set color (color: Vec3) {
        this._color.set(color);
        if (JSB) {
            this._nativeObj!.setColor(color);
        }
    }

    get color (): Vec3 {
        return this._color;
    }

    set useColorTemperature (enable: boolean) {
        this._useColorTemperature = enable;
        if (JSB) {
            this._nativeObj!.setUseColorTemperature(enable);
        }
    }

    get useColorTemperature (): boolean {
        return this._useColorTemperature;
    }

    set colorTemperature (val: number) {
        this._colorTemp = val;
        ColorTemperatureToRGB(this._colorTempRGB, this._colorTemp);
        if (JSB) {
            this._nativeObj!.setColorTemperatureRGB(this._colorTempRGB);
        }
    }

    get colorTemperature (): number {
        return this._colorTemp;
    }

    get colorTemperatureRGB (): Vec3 {
        return this._colorTempRGB;
    }

    set node (n) {
        this._node = n;
        if (this._node) {
            this._node.hasChangedFlags |= TransformBit.ROTATION;
            if (JSB) {
                this._nativeObj!.setNode(n ? n.native : null);
            }
        }
    }

    get node () {
        return this._node;
    }

    get type () : LightType {
        return this._type;
    }

    get name () {
        return this._name;
    }

    set name (n) {
        this._name = n;
    }

    get scene () {
        return this._scene;
    }

    get native (): NativeLight {
        return this._nativeObj!;
    }

    protected _baked = false;

    protected _color: Vec3 = new Vec3(1, 1, 1);

    protected _colorTemp = 6550.0;

    protected _colorTempRGB: Vec3 = new Vec3(1, 1, 1);

    protected _scene: RenderScene | null = null;

    protected _node: Node | null = null;

    protected _name: string | null = null;

    protected _useColorTemperature = false;

    protected _type: LightType = LightType.UNKNOWN;

    public initialize () {
        this._init();
        this.color = new Vec3(1, 1, 1);
        this.colorTemperature = 6550.0;
    }

    public attachToScene (scene: RenderScene) {
        this._scene = scene;
    }

    public detachFromScene () {
        this._scene = null;
    }

    public destroy () {
        this._name = null;
        this._node = null;
        this._destroy();
    }

    public update () {}
}
